### ISSUE TITLE
Could not load type 'XXX' from assembly 'Microsoft.JavaScript.No deApi.Generator.SymbolExtensions_1 exception

### ISSUE THREAD

**Issue by alexeybut** [microsoft/node-api-dotnet#201]
'TypeLoadException : Could not load type' error occurred while publishing project with AOT.
Error:
```
CSC : error NAPI1001: TypeLoadException : Could not load type 'ConsoleApp1.Class1' from assembly 'Microsoft.JavaScript.No deApi.Generator.SymbolExtensions_1, Version=0.0.0.0, Culture=neutral, PublicKeyToken=null'.   at System.Signature.GetSign ature(Void* pCorSig, Int32 cCorSig, RuntimeFieldHandleInternal fieldHandle, IRuntimeMethodInfo methodHandle, RuntimeType  declaringType)   at System.Reflection.RuntimeConstructorInfo.<get_Signature>g__LazyCreateSignature|21_0()   at System.Ref lection.RuntimeConstructorInfo.GetParametersNoCopy()   at System.RuntimeType.FilterApplyMethodBase(MethodBase methodBase,  BindingFlags methodFlags, BindingFlags bindingFlags, CallingConventions callConv, Type[] argumentTypes)   at System.Runt imeType.GetConstructorCandidates(String name, BindingFlags bindingAttr, CallingConventions callConv, Type[] types, Boolea n allowPrefixLookup)   at System.RuntimeType.GetConstructorImpl(BindingFlags bindingAttr, Binder binder, CallingConventio ns callConvention, Type[] types, ParameterModifier[] modifiers)   at Microsoft.JavaScript.NodeApi.Generator.SymbolExtensi ons.AsConstructorInfo(IMethodSymbol methodSymbol)   at System.Linq.Enumerable.WhereSelectEnumerableIterator`2.ToArray()    at Microsoft.JavaScript.NodeApi.Generator.ModuleGenerator.ExportType(SourceBuilder& s, ITypeSymbol type, String exportNa me)   at Microsoft.JavaScript.NodeApi.Generator.ModuleGenerator.ExportModule(SourceBuilder& s, ITypeSymbol moduleType, IE numerable`1 exportItems) [X:\NodeJS\ConsoleApp1\ConsoleApp1.csproj]  
```
Command:
```
dotnet publish -c DEBUG
```
Please find the sample project attached.
How to reproduce:
Unzip the attached project, run build.cmd.

net8, NodeApi 0.4.27, Windows 10 x64.
[ConsoleApp1.zip](https://github.com/microsoft/node-api-dotnet/files/14167052/ConsoleApp1.zip)

**Comment by jasongin** [comment]
Thanks. From debugging I can see this is a problem processing certain constructor and method overloads, where at least one of the overloads has a parameter that references another type that has not yet been loaded.

The order is significant; e.g. if you swap the order of the 2 overloads, the error does not occur.
```C#
        public Class2(Class1 s1) { }
        public Class2() { }
```

But the constructor overload isn't going to work correctly anyway. (JS will only be able use one of them.) As I explained in #199, overloads are not meant to be supported with `[JSExport]`. But again, the build should report a better error message. Apparently we need some negative tests for that.

**Comment by james-hu** [comment]
I am having the same issue. Is there a way we can turn marshaller generation off while keeping .d.ts generation? For example, an  entry in csprj?

**Comment by jasongin** [comment]
> Is there a way we can turn marshaller generation off while keeping .d.ts generation?

If there are no types or methods tagged with `[JSExport]`, then neither marshalling code nor a module entrypoint will be generated. In that case you can still generate typedefs and still write JS code to _dynamically_ load .NET assemblies and call APIs in them. Then the marshalling code will be generated at runtime, using reflection + code-generation -- that is less performant than the compile-time code-generation but for many applications the difference isn't significant.

**Comment by jasongin** [comment]
@james-hu Did you also have `[JSExport]` on a class with overloaded constructors or methods?

**Comment by james-hu** [comment]
I don't see overloaded constructor on the class having [JSExport]. 
There are a lot of other classes I didn't check. Those classes don't have [JSExport]

**Comment by james-hu** [comment]
reflection + code-generation works for string return type, however it doesn't work for byte[] return type. And it doesn't generate type in .d.ts for Memory<byte> return type.

**Comment by james-hu** [comment]
I don't know what are supposed to work and what are not, among byte[], IList, Memory. I tried all of them and none of them worked. My Sdk and runtime were both version 6.

**Comment by jasongin** [comment]
> There are a lot of other classes I didn't check. Those classes don't have [JSExport]

Dependency types (parameter types, return types, property types) of the exported class have to be exported also. If any of those has overloaded constructor or method then they could hit this bug. I'll try to fix it soon.

> I don't know what are supposed to work and what are not, among byte[], IList, Memory

C# `byte[]` should get marshalled as JS `Array<number>` aka `number[]`. Alternatively, C# `Memory<byte>` becomes a JS `UInt8Array`, which may be more efficient for a large buffer. Docs on this aren't great at the moment, but there is some info at https://github.com/microsoft/node-api-dotnet/blob/main/Docs/typescript.md#array-types, and a bit of [test code here](https://github.com/microsoft/node-api-dotnet/blob/975ffe9d978acba53604cd80dc662a65182dbab5/test/TestCases/napi-dotnet/complex_types.js#L77). If that isn't working how you expect, feel free to open another issue.

**Comment by james-hu** [comment]
Thanks for the explanation, I have #206 opened for the byte[] issue. I'll try to create an exported Class as return type too.

**Comment by jasongin** [comment]
@NimbusFox In any of your exported classes, is there an overloaded method that has a parameter or return type `Component`? If so the issue should be fixed by PR #221.

**Comment by NimbusFox** [comment]
Looking through my code there is no reference to System.ComponentModel.Component. I have triple checked all my exports and not one returns a component type. It's either basic structs, enums or base types. It wouldn't go looking in other libraries that my library relies on would it?

**Comment by jasongin** [comment]
> It wouldn't go looking in other libraries that my library relies on would it?

Yes, it recursively follows references to other types in any public properties, methods, or constructors.

PR #221 is not checked in yet (waiting on review from @vmoroz) so the fix is not in the 0.5.4 release.
